### ISSUE TITLE
" push a copy of input "

### ISSUE THREAD

**Issue by ricea** [whatwg/encoding#92]
https://encoding.spec.whatwg.org/commit-snapshots/a46a0c41dc471747c9bc931dfa16934c04153f98/#interface-textdecoder

>  push a copy of input 

I don't understand why this is "a copy of" rather than "a reference to". Would it be possible to clarify?

**Comment by annevk** [comment]
```javascript
const buf = new ArrayBuffer(2),
      view = new Uint8Array(buf),
      buf2 = new ArrayBuffer(1),
      view2 = new Uint8Array(buf2),
      decoder = new TextDecoder("utf-8", { ignoreBOM:true })
view[0] = 0xEF;
view[1] = 0xBB;
view2[0] = 0xBF;
assert(decoder.decode(buf, {stream:true}) === "")
assert(decoder.decode(buf2) === "\uFEFF")
```
Now if you put something like `view[0] = 0x00;` between the asserts, that should not change the outcome. The alternative would be to detach the buffer I suppose. We could maybe make that an option. @inexorabletash?

**Comment by annevk** [comment]
(We should probably add tests since Firefox doesn't pass the above rather trivial test...)

**Comment by ricea** [comment]
How about a note along the lines of:

> Implementations may elide the copy if they can guarantee that changes to _input_ after the call to `decode()` will not affect the result of future calls to `decode()`.

?

**Comment by annevk** [comment]
Sure, that's reasonable. Might even encourage better implementations. Anything that is black-box "undetectable" (performance doesn't count) is fair game.

I wonder if we should also add an API that guarantees detachment and no copy, but probably best to leave that for now.

**Comment by ricea** [comment]
I think if we had a detach API that would encourage implementations to use this "hold a reference to the original buffer" implementation style, which just seems like a bad idea:

```javascript
doSomethingWithText(decoder.decode(huge200MBbuffer, {stream: true}));
await goAwayAndHaveACupOfCoffee();
```

Backtracking to the start of the character when it is split may be a good implementation strategy, but in the worst case you just have to copy 3 bytes into scratch space to avoid holding the buffer around.

**Comment by annevk** [comment]
Yes, indeed. This is actually giving me pause about our efforts to put this kind of copying behavior in the IDL layer. We should probably more carefully consider what the right strategy is in each case.

**Comment by inexorabletash** [comment]
I assumed implementations didn't actually copy the bytes but did a synchronous decode and retain decoder state deep in the bowels of the decoder implementation (e.g. ICU, etc) as @ricea mentions; implicitly a "scratch buffer" but not necessarily coded as such. I know Blink does this, both for ICU and custom decoders. 

Since it's always that case that undetectable differences cannot preclude implementations, I don't see why it's worth specifically calling out here. Then again, if we have evidence this is an implementation mistake (is that why Gecko is failing?) no objection to a non-normative implementation hint.

And re: detaching the buffer: as an anecdote, we recently made parts of Blob creation unobservably async in Blink, but it can lead to cases where script loses a natural throttle and starts chewing up more memory faster than the rest of the system can keep up, eventually hitting OOM. The actual script-referenced memory is low, it's the implicit references to the original buffer that are bloating things. Stick ricea@'s example in a loop and the same thing will happen. So... yeah, color me nervous about blithely introducing detaching.

**Comment by inexorabletash** [comment]
See also https://github.com/whatwg/encoding/issues/44  re: implementation notes about saving state between calls.

**Comment by annevk** [comment]
The Firefox issue is probably https://bugzilla.mozilla.org/show_bug.cgi?id=1102679.

**Comment by annevk** [comment]
I ended up finding a WebKit bug as well after I modified the test: https://bugs.webkit.org/show_bug.cgi?id=169858. In any event, I didn't discover any implementation that actually did something weird with the buffer. All BOM stuff that went wrong.

**Comment by annevk** [comment]
Oh also, there's a PR that fixes this issue now: #99. Review appreciated.
